solana/indexer/subscription.go

```diff
@@ -0,0 +1,219 @@
+package indexer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	pb "github.com/rpcpool/yellowstone-grpc/examples/golang/proto"
+	"go.uber.org/zap"
+)
+
+// LaserStream from Helius only keeps the last 3000 slots.
+// Subtract 10 slots to be sure that the subscription doesn't fail.
+var MAX_SLOT_GAP = uint64(2990)
+
+type artistCoinsChangedNotification struct {
+	Operation string `json:"operation"`
+	NewMint   string `json:"new_mint"`
+	OldMint   string `json:"old_mint"`
+}
+
+func (s *SolanaIndexer) Subscribe(ctx context.Context) error {
+	conn, err := s.pool.Acquire(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to acquire database connection: %w", err)
+	}
+	defer conn.Release()
+
+	rawConn := conn.Conn()
+	_, err = rawConn.Exec(ctx, `LISTEN artist_coins_changed`)
+	if err != nil {
+		return fmt.Errorf("failed to listen for artist coins changes: %w", err)
+	}
+
+	defer func() {
+		s.logger.Info("received shutdown signal, stopping subscription")
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		coins, err := getArtistCoins(ctx, s.pool)
```

Comment:
Just curious if you have any concerns about how this scales if we have hundreds or thousands of artist coins.

Reply:
yeah this definitely will have some scaling problems going the launchpad route. I think in that world we'd probably want to shard the indexing across different subscriptions or even pods but I don't want to think about that _just_ yet